jsk_perception/src/histogram_of_oriented_gradients.cpp

```diff
@@ -1,99 +1,112 @@
+// Copyright (C) 2016 by Krishneel Chaudhary @ JSK Lab, The University
+// of Tokyo
 
 #include <jsk_perception/histogram_of_oriented_gradients.h>
 
-HOGFeatureDescriptor::HOGFeatureDescriptor() {
-   
+HOGFeatureDescriptor::HOGFeatureDescriptor(
+    const int cell_size, const int block_per_cell,
+    const int n_bins, const float angle, const int num_threads) :
+    CELL(cell_size),
+    BLOCK(block_per_cell),
+    ANGLE(angle),
+    N_BINS(n_bins),
+    num_threads_(num_threads_) {
+    this->BINS_ANGLE = this->ANGLE / this->N_BINS;
 }
 
-void HOGFeatureDescriptor::bilinearBinVoting(
+void HOGFeatureDescriptor::histogramBinVoting(
     const float &angle, int &lower_index, int &higher_index) {
     float nearest_lower = FLT_MAX;
     float nearest_higher = FLT_MAX;
     lower_index = 0;
     higher_index = 0;
     for (int i = BINS_ANGLE/2; i < ANGLE; i += BINS_ANGLE) {
-       float distance = abs(angle - i);
+	float distance = abs(angle - i);
         if (i < angle) {
             if (distance < nearest_lower) {
                 nearest_lower = distance;
                 lower_index = i;
             }
-        } else {
-            if (distance < nearest_higher) {
-                nearest_higher = distance;
-                higher_index = i;
-            }
         }
+	/*
```

Comment:
The idea is to bi-linearly interpolate, so if some distribution goes into lower bin, then the upper bin is just lower_bin+1. Save  searching time.